File: src/leetcode/java/KthSmallestElementInABST_230.java
package leetcode.java;

public class KthSmallestElementInABST_230 {
  private int cnt = 0;
  private int val = 0;
  private int tar;

  public int kthSmallest(TreeNode root, int k) {
    tar = k;
    preOrder(root);
    return val;
  }

  private void preOrder(TreeNode root) {
    if (root == null) return;

    preOrder(root.left);
    cnt++;
    if (cnt == tar) val = root.val;
    preOrder(root.right);
  }

  class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;

    TreeNode(int x) {
      val = x;
    }
  }
}

File: src/leetcode/python/KthSmallestElementInABST_230.py
class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class KthSmallestElementInABST_230(object):
    def kthSmallest(self, root, k):
        """
        :type root: TreeNode
        :type k: int
        :rtype: int
        """
        def gen(node):
            if node is not None:
                for val in gen(node.left):
                    yield val
                yield node.val
                for val in gen(node.right):
                    yield val
        for val in gen(root):
            if k == 1:
                return val
            else:
                k -= 1
